### ISSUE TITLE
[BUG] dovecot: imap,  Error: link, failed: Permission denied

### ISSUE THREAD

**Issue by melbcycle** [docker-mailserver/docker-mailserver#2053]
# Bug Report
Firstly the impact of this bug is very low. Everything seems to be working OK. This issue is being reported in the mail.err log file.
The issue relates to moving a dovecot.index.log -> dovecot.index.log.2

## Context
The docker-mailserver container is running under Portainer supervision in Docker Swarm mode. All swarm nodes have a common storage system which is NFS.

Mail.err is reporting a Permission denied error. However, the index log concerned just keeps getting bigger. Other than the size there is no other problem or effect. Everything seems OK so far (only been online a couple of days).

### What is affected by this bug?
As far as I can tell, so far, there is no impact from this issue. Only folders with lots of emails are affected (reporting an error in the mail.err file).

### When does this occur?
The error is reported when the dovecot imap indexer thinks its log has too much stuff in it. It seems to want to rotate the log file. It cannot and there are no .2 files present.

### How do we replicate the issue?

1. Run up the container with the storage connected to an NFS backend.
2. Wait for a mailbox to grow large.
3. Note the occasional errors in mail.err
4.

## Behavior
I believe it is expected that a dovecot.index.log.2 should appear with the old log information present.

### Actual Behavior
dovecot.index.log.2 never appears.

### Expected Behavior
dovecot.index.log.2 appears with old information in it.

## Your Environment
RPi48GB Docker Swarm/Portainer. NFS connected storage. Limited to 2GB ram.

- version: `v` stable
- available RAM: `GB` 2GB
- Docker version: `v` Docker version 20.10.7, build f0df350
- docker-compose version: `v` Portainer 2.5.1 - Docker Swarm mode.

### Environment Variables

```CFG

    environment:
      - OVERRIDE_HOSTNAME=mail21.domain1.com.au
      - ENABLE_SPAMASSASSIN=1
      - SPAMASSASSIN_SPAM_TO_INBOX=1
      - ENABLE_CLAMAV=1
      - ENABLE_FAIL2BAN=1
      - ENABLE_POSTGREY=1
      - ENABLE_MANAGESIEVE=1
      - ONE_DIR=1
      - DMS_DEBUG=1
      - POSTFIX_INET_PROTOCOLS=ipv4
      # >>> SSL LetsEncrypt
      - SSL_TYPE=letsencrypt
      # <<< SSL LetsEncrypt
      # >>> Postfix LDAP Integration
#     - ENABLE_LDAP=0
      - ENABLE_LDAP=1
      - LDAP_SERVER_HOST=host2.localdomain # your ldap container/IP/ServerName
      - LDAP_SEARCH_BASE=ou=people,dc=lan,dc=localdomain
      - LDAP_BIND_DN=cn=bind_ro_mailserver,dc=lan,dc=localdomain
      - LDAP_BIND_PW=password
      - LDAP_QUERY_FILTER_DOMAIN=(|(mail=*@%s)(mailAlias=*@%s)(mailGroupMember=*@%s))
      - LDAP_QUERY_FILTER_USER=(&(objectClass=inetOrgPerson)(mail=%s))
      - LDAP_QUERY_FILTER_ALIAS=(&(objectClass=inetOrgPerson)(mailAlias=%s))
      - LDAP_QUERY_FILTER_GROUP=(&(objectClass=inetOrgPerson)(mailGroupMember=%s))
      - LDAP_QUERY_FILTER_SENDERS=(&(objectClass=inetOrgPerson)(|(mail=%s)(mailAlias=%s)(mailGroupMember=%s)))
      - SPOOF_PROTECTION=1
      # <<< Postfix LDAP Integration
      # >>> Dovecot LDAP Integration
        # Filter for user lookup. Some variables can be used (see
        # http://wiki2.dovecot.org/Variables for full list):
        #   %u - username
        #   %n - user part in user@domain, same as %u if there's no domain
        #   %d - domain part in user@domain, empty if user there's no domain
      - DOVECOT_PASS_FILTER=(&(objectClass=inetOrgPerson)(mail=%u))
      - DOVECOT_PASS_ATTRS=mail=user,userPassword=password
      - DOVECOT_USER_FILTER=(&(objectClass=inetOrgPerson)(mail=%u))
        # User attributes are given in LDAP-name=dovecot-internal-name list. The
        # internal names are:
        #   uid - System UID
        #   gid - System GID
        #   home - Home directory
        #   mail - Mail location
      - DOVECOT_USER_ATTRS==home=/var/mail/%d/%n,=uid=5000,=gid=5000
      - DOVECOT_AUTH_BIND=yes
      # <<< Dovecot LDAP Integration
      # >>> SASL LDAP Authentication
#     - ENABLE_SASLAUTHD=0
      - ENABLE_SASLAUTHD=1
      - SASLAUTHD_MECHANISMS=ldap
      - SASLAUTHD_LDAP_SERVER=host12.localdomain
      - SASLAUTHD_LDAP_BIND_DN=cn=bind_ro_mailserver,dc=lan,dc=localdomain
      - SASLAUTHD_LDAP_PASSWORD=password
      - SASLAUTHD_LDAP_SEARCH_BASE=ou=people,dc=lan,dc=localdomain
#     - SASLAUTHD_LDAP_FILTER=(&(|(mail=%U@domain1.com.au)(mail=%U@domain2.com.au))(objectClass=inetOrgPerson))
#     - SASLAUTHD_LDAP_FILTER=(&(mail=%U@domain1.com.au)(objectClass=inetOrgPerson))
      - SASLAUTHD_LDAP_FILTER=(&(uid=%U)(objectClass=inetOrgPerson))
      # <<< SASL LDAP Authentication
      - POSTMASTER_ADDRESS=postmaster@domain1.com.au
      - POSTFIX_MESSAGE_SIZE_LIMIT=100000000

```

### Relevant Stack Traces

```Jun 22 07:57:28 717e51c20f1a dovecot: imap(user1@domain1.com.au)<7543><Lu138VXFjrrAqAAt>: Error: link(/var/mail/domain1.com.au/user1/.OldServer.BackupPClist/dovecot.index.log, /var/mail/domain1.com.au/user1/.OldServer.BackupPClist/dovecot.index.log.2) failed: Permission denied
Jun 22 21:13:20 717e51c20f1a dovecot: imap(user1@domain1.com.au)<46844><bGebO2HFitLAqAAt>: Error: link(/var/mail/domain1.com.au/user1/.OldServer.Uber/dovecot.index.log, /var/mail/domain1.com.au/user1/.OldServer.Uber/dovecot.index.log.2) failed: Permission denied
Jun 23 00:22:50 717e51c20f1a dovecot: imap(user1@domain1.com.au)<46844><bGebO2HFitLAqAAt>: Error: link(/var/mail/domain1.com.au/user1/.OldServer.Sent Messages/dovecot.index.log, /var/mail/domain1.com.au/user1/.OldServer.Sent Messages/dovecot.index.log.2) failed: Permission denied
Jun 23 00:28:13 717e51c20f1a dovecot: imap(user1@domain1.com.au)<46844><bGebO2HFitLAqAAt>: Error: link(/var/mail/domain1.com.au/user1/.OldServer.RedHat/dovecot.index.log, /var/mail/domain1.com.au/user1/.OldServer.RedHat/dovecot.index.log.2) failed: Permission denied
Jun 23 01:08:40 717e51c20f1a dovecot: imap(user1@domain1.com.au)<46844><bGebO2HFitLAqAAt>: Error: link(/var/mail/domain1.com.au/user1/.OldServer.Apple/dovecot.index.log, /var/mail/domain1.com.au/user1/.OldServer.Apple/dovecot.index.log.2) failed: Permission denied
Jun 23 01:17:19 717e51c20f1a dovecot: imap(user1@domain1.com.au)<46844><bGebO2HFitLAqAAt>: Error: link(/var/mail/domain1.com.au/user1/.OldServer.Amazon/dovecot.index.log, /var/mail/domain1.com.au/user1/.OldServer.Amazon/dovecot.index.log.2) failed: Permission denied
Jun 23 01:38:01 717e51c20f1a dovecot: imap(testuser@domain1.com.au)<60024><XRjr9GTFWKPAqAAt>: Error: link(/var/mail/domain1.com.au/testuser/.Trash.RedHat/dovecot.index.log, /var/mail/domain1.com.au/testuser/.Trash.RedHat/dovecot.index.log.2) failed: Permission denied

```

**Comment by casperklein** [comment]
Disclaimer: I am not familiar with docker-swarm. As the error message suggest, there is a permission problem --> with your mounts?. Maybe read-only?

**Comment by melbcycle** [comment]
Thanks Casper,

I'm not sure it is even feasible that anything would work at all if the mount was read only ? Nevertheless I have indeed verified that the user can write to the said directory. Also the log file is actually being written correctly in the same directory, it just cannot be rotated when it becomes large.

I'm pretty sure this would be the same issue even if swarm mode was not enabled, however I have not confirmed that. If so then essentially that means anyone running docker with an NFS connected mount would have this issue with the container. That also goes for Kubernetes/OpenShift. It might also effect non-dockerized installs.

Won't fix is OK too.

Just reporting it FYI. It is a very low impact problem. The mail system is working extremely well and I like it a lot.

Cheers.

**Comment by casperklein** [comment]
I think you are right.

> I'm pretty sure this would be the same issue even if swarm mode was not enabled, however I have not confirmed that.

It would be nice if you could report back your results.

**Comment by georglauterbach** [comment]
TBH, I too think that this looks like a permission error caused by volume mounts. But it does not need to be connected to Swarm.

How did you verify that permissions are correct?

**Comment by melbcycle** [comment]
Hi George,

Initially I actually did the echo "test" > test.txt thing. I was going to provide a transcript of the process and list all the other files in the directory. However.......

All these directories have been legitimately deleted. The mail folders were the subject of deletes. It seems there may be some sort of race condition. This possibly explains why there is no impact on the operation of the system as a whole.

I now suspect there is nothing wrong. Nothing being wrong usually take a long time to diagnose. I should have checked this before posting although in my defence I did do the echo test. My apologies. And thankyou for your efforts.
